### ISSUE TITLE
Second Cache Level is not working correctly with "indexBy" on association

### ISSUE THREAD

**Issue by orfin** [doctrine/doctrine2#5889]
Hi,

when using indexBy on OneToMany association, returned collection is not indexed by given field.

* getTranslations with 2ndlevel:  ```[0 => Translation, 1 => Translation]```
* getTranslations without 2ndlevel:   ```['en' => Translation, 'pl' => Translation]```

Sample mapping:

```
    /**
     * @ORM\Cache(usage="NONSTRICT_READ_WRITE", region="cms")
     * 
     * @ORM\OneToMany(targetEntity="Translation", mappedBy="page", cascade={"persist", "remove"}, indexBy="language", orphanRemoval=true)
     */
    protected $translations;


    /**
     * Get translations
     *
     * @return \Doctrine\Common\Collections\Collection
     */
    public function getTranslations()
    {
        return $this->translations;
    }
```

**Comment by lcobucci** [comment]
@Ocramius as per @FabioBatSilva's review on https://github.com/doctrine/doctrine2/pull/5899#issuecomment-228787124 I'd say we close this and improve the documentation, what do you think?

**Comment by orfin** [comment]
> I don't think that adding indexes to collections is something we want to support.

We have many projects with a vast number of associations with indexBy attribute. That is cool feature of doctrine, because it improves performance on PHP-level - we don't have to iterate over a whole array to get an element.

Removing ability to use indexBy with 2nd level cache makes this mechanizm totally unusable in our applications. We had to return to old-style ResultCache due to this limitation but would be more than happy to return to 2nd level cache :-)

**Comment by lcobucci** [comment]
@orfin the problem here is a different thing...

The ORM assumes that you have valid state on your objects even without the persistence layer. Let's consider that everything is running in memory and use your models from #5889.

If you do:

```php
$page = new DDC5889PageModel('phpunit');
$translation1 = new DDC5889TranslationModel($page, 'phpunit-1', 'en');
$translation2 = new DDC5889TranslationModel($page, 'phpunit-2', 'pl');

$page->getTranslations()->add($translation1);
$page->getTranslations()->add($translation2);

assert($page->getTranslations()->get('en') === $translation1);
assert($page->getTranslations()->get('pl') === $translation2);
```

The script will fail since the collection doesn't know how do you want to index things in it (you didn't tell it).

On the other hand, everything works with:

```php
$page = new DDC5889PageModel('phpunit');
$translation1 = new DDC5889TranslationModel($page, 'phpunit-1', 'en');
$translation2 = new DDC5889TranslationModel($page, 'phpunit-2', 'pl');

$page->getTranslations()->set('en', $translation1);
$page->getTranslations()->set('pl', $translation2);

assert($page->getTranslations()->get('en') === $translation1);
assert($page->getTranslations()->get('pl') === $translation2);
```

What we all need to understand is that there's no magic on ORMs and that we're always responsible for the lack of consistency on our objects (which will have consequences).

`indexBy` is only used to hydrate the collections on the DB side and IMHO what you need to do in your code is change all your collections that relies on that feature to use `set()` instead of `add()`.

**Comment by orfin** [comment]
But after retrieving data from the database, I'm not making any calls for add/set methods on collection. Im relying purely on Doctrine's indexBy feature which handles this. Otherwise, I'd have to use some lifecycle event to transform collection to an indexed one, which I dont want to do 'by hand'.

**Comment by lcobucci** [comment]
@orfin indeed, however it doesn't hit the database since the data is cached when it was inserted and that's why you have that behavior.
I spoke to @Ocramius and we'll plan a way to remove this inconsistency since in a nice way since things don't work when you add L2C.

My initial thought is to apply the same behavior we have when using `orderBy` - the cache is evicted instead of updated, we'll keep you updated.

**Comment by Ocramius** [comment]
Removing from `2.6.1` milestone - can't do right now.

**Comment by vitalicwow** [comment]
Would also like to see this. Using `set()` isn't a viable solution if the data is cached on read
